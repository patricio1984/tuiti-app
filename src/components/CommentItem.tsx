import { FeedItemLayout } from './ui/FeedItemLayout';
import type { Comment } from '@/api/jSONPlaceholderAPI.schemas';

export const CommentItem: React.FC<{ comment: Comment }> = ({ comment }) => {
  return (
    <FeedItemLayout
      avatarUrl={`https://i.pravatar.cc/150?u=${comment.email}`}
      header={
        <p className="text-sm text-gray-800 dark:text-gray-200 font-medium">
          {comment.name}{' '}
          <span className="text-gray-500 text-sm ml-1">
            @{comment.email?.split('@')[0] ?? 'anonimo'}
          </span>
        </p>
      }
      content={
        <p className="text-sm text-gray-600 dark:text-gray-400 mt-1">
          {comment.body}
        </p>
      }
    />
  );
};
