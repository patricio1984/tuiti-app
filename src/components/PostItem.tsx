'use client';

import { useRouter } from 'next/navigation';
import { FeedItemLayout } from './ui/FeedItemLayout';
import type { Post } from '@/api/jSONPlaceholderAPI.schemas';

export const PostItem: React.FC<{ post: Post }> = ({ post }) => {
  const router = useRouter();

  return (
    <li
      onClick={() => router.push(`/posts/${post.id}`)}
      className="cursor-pointer list-none"
    >
      <FeedItemLayout
        avatarUrl={`https://i.pravatar.cc/150?u=${post.userId}`}
        header={
          <p className="text-sm text-gray-800 dark:text-gray-200 font-medium">
            {post.title}{' '}
            <span className="text-gray-500 text-sm ml-1">
              @usuario{post.userId}
            </span>
          </p>
        }
        content={
          <p className="text-sm text-gray-600 dark:text-gray-400 mt-1">
            {post.body}
          </p>
        }
      />
    </li>
  );
};
