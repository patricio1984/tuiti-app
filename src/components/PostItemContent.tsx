import type { Post } from '@/api/jSONPlaceholderAPI.schemas';

export const PostItemContent = ({ post }: { post: Post }) => {
  return (
    <article
      className="border border-gray-200 dark:border-neutral-700
                 rounded-xl p-4 mb-6 transition-colors duration-200"
    >
      <div className="flex items-start gap-3">
        <div className="w-10 h-10 rounded-full bg-gray-300 dark:bg-neutral-700" />
        <div className="flex-1">
          <p className="text-sm text-gray-800 dark:text-gray-200 font-medium">
            @usuario{post.userId}
          </p>
          <p className="text-base text-gray-900 dark:text-gray-100 mt-1">
            {post.title}
          </p>
          <p className="text-sm text-gray-600 dark:text-gray-400 mt-1">
            {post.body}
          </p>
        </div>
      </div>
    </article>
  );
};
