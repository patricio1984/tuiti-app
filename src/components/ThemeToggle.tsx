'use client';

import { useTheme } from '@/lib/theme';
import { Moon, Sun } from 'lucide-react';

export const ThemeToggle = () => {
  const { theme, toggleTheme } = useTheme();

  return (
    <button
      onClick={toggleTheme}
      aria-label="Cambiar tema"
      className="cursor-pointer ml-auto p-2 rounded hover:bg-gray-200 dark:hover:bg-gray-800 transition"
    >
      {theme === 'dark' ? (
        <Moon className="w-5 h-5 text-white" />
      ) : (
        <Sun className="w-5 h-5 text-[#1DA1F2]" />
      )}
    </button>
  );
};
