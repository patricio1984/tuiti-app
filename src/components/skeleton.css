@layer utilities {
  .shimmer-bg {
    background: linear-gradient(
      90deg,
      rgba(229, 229, 229, 0.1) 0%,
      rgba(229, 229, 229, 0.3) 50%,
      rgba(229, 229, 229, 0.1) 100%
    );
    background-size: 200% 100%;
    animation: shimmer 1.5s infinite linear;
  }

  @media (prefers-color-scheme: dark) {
    .shimmer-bg {
      background: linear-gradient(
        90deg,
        rgba(75, 85, 99, 0.1) 0%,
        rgba(75, 85, 99, 0.3) 50%,
        rgba(75, 85, 99, 0.1) 100%
      );
    }
  }

  @keyframes shimmer {
    0% {
      background-position: -200% 0;
    }
    100% {
      background-position: 200% 0;
    }
  }
}
