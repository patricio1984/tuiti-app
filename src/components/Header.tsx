import { TwitterLogo } from './TwitterLogo';
import { ThemeToggle } from './ThemeToggle';
import Link from 'next/link';

export const Header = () => {
  return (
    <header className="fixed top-0 inset-x-0 z-50 border-b border-gray-200 dark:border-gray-800 bg-white/80 dark:bg-black/80 backdrop-blur supports-[backdrop-filter]:bg-white/60 dark:supports-[backdrop-filter]:bg-black/60">
      <div className="flex items-center justify-between max-w-2xl mx-auto px-4 h-14">
        <Link href="/" aria-label="Inicio" className="flex items-center gap-2">
          <TwitterLogo />
        </Link>

        <ThemeToggle />
      </div>

      <h1 className="sr-only">Aplicaci√≥n de Posts - JSONPlaceholder</h1>
    </header>
  );
};
