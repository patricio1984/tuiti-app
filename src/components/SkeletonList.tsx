import React from 'react';
import './skeleton.css';

export const SkeletonList: React.FC<{ count: number }> = ({ count }) => {
  return (
    <ul className="divide-y divide-gray-200 dark:divide-gray-700">
      {Array.from({ length: count }).map((_, i) => (
        <li
          key={i}
          className="px-4 py-4 hover:bg-gray-100 dark:hover:bg-gray-800 transition"
        >
          <div className="flex items-start space-x-4">
            <div className="w-10 h-10 rounded-full shimmer-bg" />
            <div className="space-y-2 flex-1">
              <div className="h-4 w-1/3 shimmer-bg rounded" />
              <div className="h-3 w-full shimmer-bg rounded" />
              <div className="h-3 w-5/6 shimmer-bg rounded" />
            </div>
          </div>
        </li>
      ))}
    </ul>
  );
};
