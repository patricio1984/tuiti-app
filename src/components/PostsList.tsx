'use client';

import React, { useRef, useEffect, useMemo } from 'react';
import { useInfinitePosts } from '@/hooks/useInfinitePosts';
import { PostItem } from './PostItem';
import { SkeletonList } from './SkeletonList';

export const PostsList: React.FC = () => {
  const {
    data,
    fetchNextPage,
    hasNextPage,
    isFetchingNextPage,
    isLoading,
    isError,
  } = useInfinitePosts();

  const loadMoreRef = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    const observer = new IntersectionObserver(
      (entries) => {
        if (entries[0].isIntersecting && hasNextPage) {
          fetchNextPage();
        }
      },
      { rootMargin: '300px' }
    );

    if (loadMoreRef.current) observer.observe(loadMoreRef.current);
    return () => observer.disconnect();
  }, [fetchNextPage, hasNextPage]);

  const shuffledPosts = useMemo(() => {
    const allPosts = data?.pages.flatMap((page) => page) ?? [];
    return allPosts.sort(() => Math.random() - 0.5);
  }, [data]);

  return (
    <div className="max-w-2xl mx-auto bg-white dark:bg-black min-h-screen">
      <ul>
        {isLoading ? (
          <SkeletonList count={10} />
        ) : isError ? (
          <li className="text-center text-red-500 py-4">
            ‚ùå Error al cargar los posts.
          </li>
        ) : (
          shuffledPosts.map((post) => <PostItem key={post.id} post={post} />)
        )}
      </ul>

      {isFetchingNextPage && <SkeletonList count={3} />}
      <div ref={loadMoreRef} className="h-1" />
    </div>
  );
};
