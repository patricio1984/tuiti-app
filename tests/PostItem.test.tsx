import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { PostItem } from '@/components/PostItem';
import type { Post } from '@/api/jSONPlaceholderAPI.schemas';
import { useRouter } from 'next/navigation';
import { vi, type Mock } from 'vitest';

vi.mock('next/navigation', () => ({
  useRouter: vi.fn(),
}));

describe('PostItem', () => {
  const post: Post = {
    id: 1,
    title: 'Post de prueba',
    body: 'Contenido del post',
    userId: 1,
  };

  beforeEach(() => {
    (useRouter as Mock).mockReturnValue({
      push: vi.fn(),
    });
  });

  it('renderiza título y contenido', () => {
    render(<PostItem post={post} />);
    expect(screen.getByText(/Post de prueba/i)).toBeInTheDocument();
    expect(screen.getByText(/Contenido del post/i)).toBeInTheDocument();
  });

  it('navega al hacer clic en el post', () => {
    const mockPush = vi.fn();
    (useRouter as Mock).mockReturnValue({ push: mockPush });

    render(<PostItem post={post} />);
    fireEvent.click(screen.getByRole('listitem'));
    expect(mockPush).toHaveBeenCalledWith('/posts/1');
  });

  it('debería tener role="listitem"', () => {
    render(<PostItem post={post} />);
    const listItem = screen.getByRole('listitem');
    expect(listItem).toBeInTheDocument();
  });

  it('renderiza correctamente (snapshot)', () => {
    const { container } = render(<PostItem post={post} />);
    expect(container).toMatchSnapshot();
  });
});
