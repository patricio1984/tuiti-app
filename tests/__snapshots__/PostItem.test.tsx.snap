// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostItem > renderiza correctamente (snapshot) 1`] = `
<div>
  <li
    class="cursor-pointer list-none"
  >
    <div
      class="flex items-start gap-3 px-4 py-3 border border-gray-200 dark:border-neutral-700
                 -mb-px animate-fade-in hover:bg-gray-100 dark:hover:bg-[#16181c] transition-colors duration-200"
    >
      <img
        alt=""
        class="w-10 h-10 rounded-full bg-gray-300 dark:bg-neutral-700 object-cover"
        src="https://i.pravatar.cc/150?u=1"
      />
      <div
        class="flex-1"
      >
        <p
          class="text-sm text-gray-800 dark:text-gray-200 font-medium"
        >
          Post de prueba
           
          <span
            class="text-gray-500 text-sm ml-1"
          >
            @usuario
            1
          </span>
        </p>
        <div
          class="mt-1"
        >
          <p
            class="text-sm text-gray-600 dark:text-gray-400 mt-1"
          >
            Contenido del post
          </p>
        </div>
      </div>
    </div>
  </li>
</div>
`;
